interpreter/cling/lib/Utils/ASTImportSource.cpp

```diff
@@ -0,0 +1,189 @@
+        return false;
+
+      /* Prepare to import the Decl(Context)  we found in the
+       * child interpreter */
+      const FileSystemOptions systemOptions;
+      FileManager fm(systemOptions, nullptr);
+
+      /*****************************ASTImporter**********************************/
+      ASTImporter importer(to_ASTContext, fm, from_ASTContext, fm,
+        /*MinimalImport : ON*/ true);
+      /**************************************************************************/
+
+      /* If this Name we are looking for is for example a Namespace,
+       * then it is a Decl Context. */
+      if ((*lookup_result.data())->getKind() == Decl::Namespace) {
+        DeclContext *declContextToImport =
+          llvm::cast<DeclContext>(*lookup_result.data());
+        ImportDeclContext(declContextToImport, importer, childDeclName,
+                          parentDeclName, childCurrentDeclContext);
+      } else { /* If this name is just a Decl. */
+        /* Check if we have more than one results, this means that it
+         * may be an overloaded function. */
+        if (lookup_result.size() > 1) {
+          DeclContext::lookup_iterator I;
+          DeclContext::lookup_iterator E;
+          for (I = lookup_result.begin(), E = lookup_result.end(); I != E; ++I) {
+            NamedDecl *D = *I;
+            Decl *declToImport = llvm::cast<Decl>(D);
+            ImportDecl(declToImport, importer, childDeclName,
+                       parentDeclName, childCurrentDeclContext);
+          }
+        } else {
+          Decl *declToImport = llvm::cast<Decl>(*lookup_result.data());
+          ImportDecl(declToImport, importer, childDeclName,
+                     parentDeclName, childCurrentDeclContext);
+        }
+      }
+      return true;
+    }
+
+    ///\brief This is the most important function of the class ASTImportSource
+    /// since from here initiates the lookup and import part of the missing
+    /// Decl(s) (Contexts).
+    ///
+    bool ASTImportSource::FindExternalVisibleDeclsByName(
+      const DeclContext *childCurrentDeclContext, DeclarationName childDeclName) {
+
+      assert(childCurrentDeclContext->hasExternalVisibleStorage() &&
+             "DeclContext has no visible decls in storage");
+
+      /* clang will call FindExternalVisibleDeclsByName with an
+       * IdentifierInfo valid for the child interpreter. Get the
+       * IdentifierInfo's StringRef representation.
+       * Get the identifier info from the parent interpreter
+       * for this Name. */
+      llvm::StringRef name(childDeclName.getAsString());
+      IdentifierTable &parentIdentifierTable =
+        m_parent_Interp->getCI()->getASTContext().Idents;
+      IdentifierInfo &parentIdentifierInfo = parentIdentifierTable.get(name);
+      DeclarationName parentDeclName(&parentIdentifierInfo);
+
+      /* Check if we have already imported this Decl (Context). */
+      if (m_DeclName_map.find(childDeclName.getAsString()) != m_DeclName_map.end())
```

Comment:
Why do you first lookup the parent name and then check whether we need it? I'd reverse these two...